### PR TITLE
Allow more immutable funcs for RUF009

### PR DESCRIPTION
**Pull request by qdegraaf**
## Summary

`RUF009` wrongly flags on `float("-inf")` calls. As discussed in https://github.com/charliermarsh/ruff/issues/4657 we can take this moment to add more funcs with immutable results.

TODOs:

- [x] Check for completeness of added funcs
- [x] @charliermarsh to confirm that we are sure we want to allow these funcs in `flake8_bugbear` as well (as discussed in the issue). The tests for bugbear seem quite explicit about only allowing certain `float()` calls for example. If changed Ruff would appear to deviate from upstream implementation. Which might be desirable in this case, but if not, PR can be changed to add a separate constant for `RUF009`.

## Test Plan

Cases added to `RUF009.py` 
Test TODOs:

- [x] Change comments in `flake8_bugbear` tests if we end up editing its logic

Closes: https://github.com/charliermarsh/ruff/issues/4657

### CODE REVIEW

**Comment by charliermarsh** [comment]
Oh, this is interesting, I didn't realize we had special-casing for those floats... How would you feel about instead changing our other rules to respect that same carveout?

**Comment by qdegraaf** [comment]
> Oh, this is interesting, I didn't realize we had special-casing for those floats... How would you feel about instead changing our other rules to respect that same carveout?

I would be happy to! Just for my understanding, and to be complete, to what degree does this apply to other primitive Python functions? The original `flake8-bugbear` rule states:

> "B008 Do not perform function calls in argument defaults.  The call is "
        "performed only once at function definition time. All calls to your "
        "function will reuse the result of that definition-time function call.  If "
        "this is intended, assign the function call to a module-level variable and "
        "use that variable as a default value."

So I take it calling `float("-inf")` or a similar infinite value avoids this necessity somehow. 

Essentially what I am wondering is: if the following default args are ok, and they do not need to be moved to module vars:

```python
def foo(x: float = float("-inf"), y: typing.Tuple[int] = tuple(1,2)):
    ...
``` 

why do these need to be moved (according to bugbear)?
```python
def bar(x: str = str("bla"), y: int = int(42)):
    ...
```

As far as I understand these primitives, they all create immutable values that are fine to be reused (or as fine as their allowed counterparts). I'm assuming it's not an oversight in bugbear or here, so I'm just checking so I can understand what the difference is and I can be certain I'm not adding or removing the wrong function to either `RUF` or `B`

**Comment by charliermarsh** [comment]
So in this case, I'd go to the source and try to find a record of why / how the decision was made...

In this case, I found https://github.com/PyCQA/flake8-bugbear/pull/155 and https://github.com/PyCQA/flake8-bugbear/issues/154. It sounds like originally, bugbear didn't allow any `float` calls (even `float("-inf")` or similar). The issue was raised to support those cases, since they don't have corresponding literal definitions, and they made the choice to be conservative, and only support cases for which there aren't corresponding literals. E.g., `x: float = float("5")` is fine, but odd (why not use a literal?), and so why allow it?

I'm fine with either decision... Having looked at the source issue, I guess I'm leaning towards allowing arbitrary `int`, `float`, etc., calls, since they _do_ produce immutable values, and if we want to suggest that users should use `5.0` instead of `float("5.0")`, that feels like it could be the responsibility of a separate rule, if needed.

**Comment by qdegraaf** [comment]
Ah nice find, thanks for the effort. Don't know why I just assumed it was a Python thing and not a bugbear design choice. Will do some more digging before throwing questions up for any future plugin port confusion. 

I agree with leaning towards allowing all immutable func calls, for the reasons you stated, as well as avoiding any similar confusion someone might have in the future. If you're happy to deviate from upstream implementation I can add some more immutable funcs, will tweak the bugbear tests and double-check both `RUF` and `B`. I can make a spinoff issue for implementing the separate rule for the `5.0` instead of `float("5.0")` etc. as well if it seems useful,

File: crates/ruff/src/rules/flake8_bugbear/snapshots/ruff__rules__flake8_bugbear__tests__B008_B006_B008.py.snap [hunk 1]
```diff
@@ -46,37 +46,6 @@ B006_B008.py:119:30: B008 Do not perform function call in argument defaults
...
     |
 
-B006_B008.py:157:34: B008 Do not perform function call `float` in argument defaults
-    |
-157 | def float_infinity_literal(value=float("1e999")):
-    |                                  ^^^^^^^^^^^^^^ B008
-158 |     pass
-    |
-
-B006_B008.py:162:30: B008 Do not perform function call `float` in argument defaults
```
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_bugbear/snapshots/ruff__rules__flake8_bugbear__tests__B008_B006_B008.py.snap:56]: We should probably update these comments, they're not out-of-date, I think? (Can we also remove the special-casing code in the bugbear module, which should now be redundant.)

Comment (qdegraaf) [reply crates/ruff/src/rules/flake8_bugbear/snapshots/ruff__rules__flake8_bugbear__tests__B008_B006_B008.py.snap:56]: I'm not sure I follow exactly on the comments. We settled on allowing `float` calls now right? So these snapshots can go. 

I updated the comments in the `.py` fixture to reflect this. As for the Violating string it's formatted like so:

```rust
format!("Do not perform function call `{name}` in argument defaults")
```

So this needs no edit. Does that cover what you mean or are you referring to something else. 

As for the special casing float logic, I'm removing it now!

Comment (charliermarsh) [reply crates/ruff/src/rules/flake8_bugbear/snapshots/ruff__rules__flake8_bugbear__tests__B008_B006_B008.py.snap:56]: Sorry, I meant "they're out-of-date", not "they're _not_ out-of-date". What you've done here looks correct to me.

Comment (qdegraaf) [reply crates/ruff/src/rules/flake8_bugbear/snapshots/ruff__rules__flake8_bugbear__tests__B008_B006_B008.py.snap:56]: Ah check! Removed infinity/nan logic now as well!

**Comment by charliermarsh** [comment]
LGTM, thanks!

**Pull request merged**